ui/app/js/controllers/permissions_controller.js

```diff
@@ -6,35 +6,73 @@ function($scope, $routeParams, $location, $timeout, Permissions, Search, $modal,
   $scope.loading = true;
   $scope.failed = false;
   $scope.username = $routeParams.username;
+  $scope.users = [];
+  $scope.tab = 1;
+  $scope.current_role = "All Roles";
+
+  $scope.isAllRoles = function (role) {
+    return $scope.current_role === "All Roles";
+  };
+
+  $scope.isSet = function (tabNum) {
+      return $scope.tab === tabNum;
+  };
+
+  $scope.setTab = function (newTab) {
+    $scope.tab = newTab;
+  };
+  $scope.roles_users = {};
+  $scope.roles_list = ["All Roles"];
+
+
   if ($scope.username) {
     // history of a specific rule
     Permissions.getUserPermissions($scope.username)
-    .then(function(response) {
-      $scope.permissions = response;
-    });
+      .then(function (response) {
+        $scope.permissions = response;
+      });
   } else {
     Permissions.getUsers()
-    .success(function(response) {
-      $scope.users = _.map(response.users, function (each) {
-        return {username: each};
+      .success(function (response) {
+        $scope.users = _.map(response.users, function (each) {
```

Comment:
Further to the above point, you should be able to use the data in the response here to build up the entire list of roles and their associated users.

Reply:
I have used the data from the `api/users` response to build the entire roles and their associated users. But I also made use of $scope.roles_users and $scope.roles_list to hold the respective data after the build. I couldn't figure out a way to do it with less. At least not yet.  I will appreciate more clarification on this. Thanks

Reply:
What I'm suggesting is that `$scope.users` should be a single data structure that contains both user and role information. Something structured like:
```
{
   "bob": {
      "roles": ["releng", "relman"]
   },
   "obinna": {
      "roles": ["relman", "qe"]
   }
}
```

This single data structure contains all of the information that your current `$scope.users`, `$scope.roles_list`, and `$scope.roles_users` currently does, which means when changes are made there's only one data structure to update.

Using this unified data structure means that code that currently uses `$scope.roles_users` (like https://github.com/mozilla/balrog/pull/417/files#diff-5c2e73a618f68d1bf0ac1eabb3d8dff8R107) will need to be adjusted to flip the data structure around. For example, you could have a function called `getRoleUsers` that looks iterates over `$scope.users` and returns a data structure where roles are the top level key.

Does that clarify things at all?

Reply:
Oh. Yeah.  Now I get. I will effect changes.